### DISCUSSION TITLE
[Hackathon Use Case] - Jupyter Notebook with FDC3 integration

### DISCUSSION THREAD

**Discussion by maoo** [finos/finos-pmcs#253] [Ideas]
# Jupyter Notebook with FDC3 integration
_From @chuck-adaptive (Adaptive)_

## Business Problem

When working with data, I need to be able to leverage cloud services for both running AI models and collaborating with teammates. This positions Jupyter Notebooks as the new excel - a user friendly GUI that unleashes powerful interactive patterns to analyze data, and even build and publish your own AI models.

Users will demand the same level of interoperability between Excel and Desktop Applications as they do between their Jupyter Notebooks - both deployed in managed services, and locally ran for quantitative development teams.

The FINOS [JupyterLab Templates](https://github.com/finos/jupyterlab_templates) are a great place to start embedding this behavior.

## Expected Features

* A Notebook template capable of leveraging FINOS hosted UI elements - i.e. `regular-table` or `Perspective`
* `FDC3` agent discovery and APIs - let the notebook discover if the Browser or Container has the agent available
* Basic Notebook UI elements for interacting with the FDC3 API
* Typings support
* Documentation for features

## Useful apps, systems, projects, technologies, libraries and datasets

- [JupyterLab Templates](https://github.com/finos/jupyterlab_templates)
- [FDC3](https://fdc3.finos.org)
- [Perspective](https://github.com/finos/perspective)
- [Regular Table](https://github.com/finos/regular-table)

**Comment by chuck-adaptive** [comment]
Perspective already has widgets - https://perspective.finos.org/docs/python.html#jupyterlab

**Comment by chuck-adaptive** [comment]
Ah, same for regular-table - https://pypi.org/project/ipyregulartable/

So we will focus on the FDC3 side of things 👍

**Comment by garrettladley** [comment]
@chuck-adaptive found your pitch very interesting! How can I help with the Python type hints?

**Comment by chuck-adaptive** [comment]
@garrettladley Terrific!

The team has done some preliminary research. In the end, what I hope to achieve with type hints is enabling Python developers to "discover" the FDC3 API via type hints, as I so enjoyed doing with Typescript. I can trust my integration points, because the types feed me, etc.

For reference - here are the Typescript types - https://github.com/finos/FDC3

There are some libraries that do this automatically

https://pypi.org/project/ts2python/

But I haven't fully evaluated them

The key thing here is to make a best effort to encapsulate the Typescript types in Python.

in my research, I liked this comparison - we want to be able to account for optional keys

```
// Typescript

interface Icon {
  src: string;
  size?: string;
  type?: string;
}
```

```
# Python

class Icon(dict):
  def __init__(self, src: string, size: str = None, type: str = None):
    super().__init__(src=src, size=size, type=type)

  @property
  def src(self) -> str:
    return self['src']

  @property
  def size(self) -> str:
    return self.get('size', '') # maybe this should be None?

  ...
```

`Ultimate Goal` - a build pipeline that consume the latest Typescript definitions and creates Python types. I say that only to say, if we see a means for automating this, that is very exciting. i.e. a new commit updates the FDC3 typings (Typescript) and this kicks off a build that updated these types

**Comment by andleebc** [comment]
@chuck-adaptive 

Hi Chuck! We are interested in this use case :) 

Could you please provide more information on what it entails? We are uncertain on where to start.

**Reply by iandilling-adaptive** [reply]
I would start by [creating custom Jupyter widgets](https://ipywidgets.readthedocs.io/en/8.0.5/examples/Widget%20Custom.html) that wrap the [fdc3 js library](https://www.npmjs.com/package/@finos/fdc3), then opening the Jupyter notebook in a fdc3 compatible desktop agent and trying out the widgets.

**Reply by chuck-adaptive** [reply]
We will sync on some widgets we've created today

**Comment by robmoffat** [comment]
Hi @chuck-adaptive,

@andleebc and colleagues would really like a chat with you about this to make sure they're on the right track - I've tried to give a bit of background about the above comment but if you could spare a 1/2 hour with them tomorrow morning that would be amazing!

cheers!

**Reply by chuck-adaptive** [reply]
💯

**Reply by chuck-adaptive** [reply]
I will be in tomorrow morning, excited to chat about the use case

Our team has assembled a few widgets already - but in my opinion there's lot more that can be done

**Comment by TheJuanAndOnly99** [comment]
Closing discussion in preparation for [this years hackathon](https://www.finos.org/hosted-events/2024-10-25-citi-india-hackathon-supported-by-finos). Feel free to reopen and update if its still relevant.
